### PR TITLE
dyno: Resolve nested functions without outer variables

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR adds support to the new compiler for resolving nested functions that do not contain any references to outer variables.

To do so, it adds a new meta-type `OuterVariables`, which stores the outer variables used in a nested function as well as their mentions. Currently, both are stored in lexical order. Note that module-scope variables are not considered to be an outer variable for the purposes of this analysis, as they have an infinite lifetime.

The set of outer variables is computed using a new query `computeOuterVariables`. This query is implemented by leveraging the `scopeResolveFunction` query, since that also necessitates full traversal of the function. Note that determination of outer variables for a given nested function 'NF' also requires us to call `computeOuterVariables` for any child functions of NF, so that we can propagate distant outer variables.

The queries `resolveFunction` and `resolveConcreteFunction` as well as `instantiateSignature` have been adjusted. If a nested function does not capture outer variables, it can be resolved using the normal process. In this case, nesting of the function is just a syntactic convenience for the user, and does not affect the function's resolution at all.

I've added a predicate `idIsNestedFunction` to the parsing queries header.

I've also introduced a new `ID` method called `isSymbolDefiningScope`, which before was implicitly encoded in the test `id.postOrderId() == -1`.

FUTURE WORK

- We may not need to store mentions in `OuterVariables`, and we may not need to store variables in lexical order. In this case, the type can be simplified. I leave that as future work.
- Nested functions that use outer variables will need to be resolved eventually, and the implementation will probably consist of a combination of: storing close outer variables (those defined in the scope of our immediate parent) in a function's `TypedFnSignature`, as well as passing in some sort of `CallContext` when resolving a call to a (potentially nested) function. The `CallContext` would include the types of outer variables represented in some form, as well as the current `PoiInfo`. This would enable us to instantiate the initial signature for a nested function and perform resolution as normal.

Reviewed by @DanilaFe. Thanks!

### CODE REVIEW

File: frontend/lib/resolution/resolution-queries.cpp [hunk 1]
```diff
@@ -2169,14 +2185,55 @@ scopeResolveFunctionQuery(Context* context, ID id) {
...
+
+  // For now, preemptively return 'nullptr' if 'id' is not a function.
+  if (!parsing::idIsNestedFunction(context, id)) return nullptr;
+
+  if (!context->hasCurrentResultForQuery(computeOuterVariablesQuery, { id })) {
+    if (!context->isQueryRunning(scopeResolveFunctionQuery, { id })) {
+
+      // The 'computeOuterVariablesQuery' is set as a side effect of
+      // performing scope resolution, since both require a traversal.
+      std::ignore = scopeResolveFunction(context, id);
```
Comment (DanilaFe) [inline frontend/lib/resolution/resolution-queries.cpp:2225]: A major question I have about this is the following: I believe long term, "scope resolution" will not exist, and be replaced by resolution. At least, this was my understanding of long trm Dyno goals. (under my understanding, scope resolution only exists to get more of the production compiler using Dyno sooner). If that were the case, will this still be workable? Will you need to run resolution to get out variables, and will you need outer variables to run resolution?

File: frontend/lib/resolution/resolution-types.cpp [hunk 2]
```diff
@@ -40,6 +40,33 @@ namespace resolution {
...
+  const bool isReachingUse = symbolParent != varParent;
+  const bool isChildUse = mentionParent != symbol_;
+
+  CHPL_ASSERT(varParent != symbol_);
+  if (!isReachingUse) {
+    CHPL_ASSERT(mention && symbol_.contains(mention));
+  }
+
+  auto it = variableToMentionIdx_.find(var);
+  auto& p = variableToMentionIdx_[var];
```
Comment (DanilaFe) [inline frontend/lib/resolution/resolution-types.cpp:56]: You can save the work of two lookups by doing an `emplace_hint` using `it`.

```C++
auto it = variableToMentionIdx_.find(var);
if (it == variableToMentionIdx_.end()) {
  // Note: emplace_hint returns an iterator to the inserted value.
  it = variableToMentionIdx_.emplace_hint(it, variables_.size(), {});
  variables.push_back(var);
  if (isReachingUse) numReachingVariables_++;
}
if (!isChildUse) {
  it->second.push_back(mentions_.size());
  mentions_.push_back(mention);
}
```

**Review by DanilaFe — APPROVED** [review]

File: frontend/include/chpl/resolution/resolution-types.h [hunk 3]
```diff
@@ -317,6 +317,167 @@ class UntypedFnSignature {
...
+  ID symbol_;
+
+  // The immediate parent of 'symbol_'. So that we can detect if a variable
+  // is 'reaching' without needing the compiler context.
+  ID parent_;
+
+  template <typename T>
+  static inline bool inBounds(const std::vector<T> v, size_t idx) {
+    return 0 <= idx && idx < v.size();
+  }
```
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply frontend/include/chpl/resolution/resolution-types.h:360]: I'm going to leave this embedded in the type for now - I'd rather add that to a common header at the same time that we make a concerted effort to adjust bounds check in other types. I'll make a private task as an IOU.

File: frontend/include/chpl/resolution/resolution-types.h [hunk 4]
```diff
@@ -317,6 +317,167 @@ class UntypedFnSignature {
...
+
+  // Mutating method used to build up state.
+  void add(Context* context, ID mention, ID var);
+
+  /** Returns 'true' if there are no outer variables. */
+  inline bool isEmpty() const { return numVariables() == 0; }
+
+  /** The total number of outer variables. */
+  inline int numVariables() const { return variables_.size(); }
```
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply frontend/include/chpl/resolution/resolution-types.h:412]: Wow I had absolutely no idea.

File: frontend/lib/parsing/parsing-queries.cpp [hunk 5]
```diff
@@ -857,6 +857,14 @@ bool idIsParenlessFunction(Context* context, ID id) {
   return idIsFunction(context, id) && idIsParenlessFunctionQuery(context, id);
 }
 
+bool idIsNestedFunction(Context* context, ID id) {
+  if (id.isEmpty() || !idIsFunction(context, id)) return false;
+  if (auto up = id.parentSymbolId(context)) {
```
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply frontend/lib/parsing/parsing-queries.cpp:862]: Yes, I think `g` is still a nested function. Even if the parent is anonymous.

[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply frontend/lib/parsing/parsing-queries.cpp:862]: Ok I think you mean the issue is that the anonymous functions do not have symbol IDs - if that is the case then they probably should.

File: frontend/lib/resolution/Resolver.cpp [hunk 6]
```diff
@@ -418,6 +420,20 @@ types::QualifiedType Resolver::typeErr(const uast::AstNode* ast,
...
+  if (target.isEmpty()) return false;
+
+  // E.g., a function, or a class/record/union/enum. We don't need to track
+  // this, and shouldn't, because its current instantiation may not make any
+  // sense in this context. As well, these things have an "infinite lifetime",
+  // and are always reachable.
+  if (target.isSymbolDefiningScope()) return false;
+  auto up = target.parentSymbolId(rv->context);
+  auto ast = parsing::idToAst(rv->context, up);
+  bool ret = ast && ast != rv->symbol && ast->isFunction();
```
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply frontend/lib/resolution/Resolver.cpp:433]: Note that I need to be more specific about the lifetime of `someVar`.

File: frontend/lib/resolution/Resolver.cpp [hunk 7]
```diff
@@ -2962,10 +2985,30 @@ bool Resolver::enter(const NamedDecl* decl) {
...
+        outerVars.get()) {
+      if (auto ovs = computeOuterVariables(context, idChild)) {
+        for (int i = 0; i < ovs->numVariables(); i++) {
+
+          // If the variable is reaching in the child function, it means it
+          // was defined in one of _our_ parent(s). So we need to track it.
+          if (ovs->isReachingVariable(i)) {
+            ID var = ovs->variable(i);
+            ID mention = ovs->firstMention(var);
+            outerVars->add(context, mention, var);
```
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply frontend/lib/resolution/Resolver.cpp:3007]: I left a comment saying that mentions are not recorded for grandchildren, so just grab the first (as a convenience).

File: frontend/lib/resolution/resolution-queries.cpp (hunk 1 above)
Comment (dlongnecke-cray) [reply frontend/lib/resolution/resolution-queries.cpp:2225]: We'll just cannibalize the scope-resolution code and use it to collect outer variables.

**Pull request merged**